caffe2/CMakeLists.txt

```diff
@@ -523,6 +523,7 @@ if (BUILD_PYTHON)
       caffe2_pybind11_state caffe2_library)
   if (WIN32)
     target_link_libraries(caffe2_pybind11_state ${PYTHON_LIBRARIES})
+    target_link_libraries(caffe2_pybind11_state onnx_proto)
```

Comment:
We don't need to link to onnx_proto. If you do `dumpbin /exports onnx_proto.lib`, then you can see that it's actually empty.

Reply:
There are two types of linking in Windows: static (against .lib) and dynamic (against .dll with import .lib library).
I don't have experience with dumpbin /exports, however it seems to be a tool to show symbols, which lib file exports for connection with dll. In that case it will not show anything for any standard static lib, that is not intended to be used as import library for dll linkage. But probably I'm mistaken here, since of lack of usage of dumpbin /exports.

Concrete reason why we link here against onnx_proto is the line:

::ONNX_NAMESPACE::ValueInfoProto vi{};

in pybind_state.cc file.

This line calls the constructor, which will be unresolved in pybind .pyd file and give us an error:

public: __cdecl onnx_c2::ValueInfoProto::ValueInfoProto(void)

Reply:
`add_library(onnx_proto ${ONNX_PROTO_SRCS} ${ONNX_PROTO_HDRS})` stays in third_party\onnx, so BUILD_SHARED_LIBS variable doesn't affect it.

Pybind is declared as

`add_library(caffe2_pybind11_state MODULE ${Caffe2_CPU_PYTHON_SRCS})`

As far as I am aware specifically in Windows (but not other platforms) MODULE and SHARED are essentially the same (dll).  And that's why `target_link_libraries` is allowed inside `if (WIN32)` for MODULE lib.
.pyd extension is used just for python import statement to work.

If .pyd is just dll, then we definitely need to link it with unresolved externals as any other dll. 

--------------

I experimented with BUILD_SHARED_LIBS variable in cmake.
Setting it to OFF results in too big caffe2.lib file in case of Debug build. Consequently, I get an error when trying to build pybind:

> D:\MyPytorch\pytorch\build\Debug\lib\Debug\caffe2.lib : fatal error LNK1248: image size (10023B2FA) exceeds maximum allowable size (FFFFFFFF) [D:\MyPytorch\pytorch\build\Debug\caff
>        e2\caffe2.vcxproj]

My cl.exe compiler is explicitly set to be x64 (as in deleted script).

It seems that for building pybind BUILD_SHARED_LIBS==ON is the only option for CMAKE_BUILD_TYPE==Debug case. In this case caffe2.dll is compiled and caffe2.lib is just an import library for that dll.

Reply:
https://github.com/pytorch/pytorch/blob/f000101b8139378f342b175c11072a925c9d7c7a/cmake/Dependencies.cmake#L896

This seems to be the exact line, where BUILD_SHARED_LIBS stops affecting onnx.

Reply:
@ArutyunovG This is what I'm fixing in #13100.